File: controllers/securityQuestionController.js
const SecurityQuestion = require('../model/securityQuestion');

// Add or update security questions for a user
exports.addOrUpdateSecurityQuestions = async (req, res) => {
  try {
    const userId = req.user._id;
    let securityQuestions =req.body; // Expecting an array of security questions and answers

    if (!Array.isArray(securityQuestions) || securityQuestions.length === 0) {
      securityQuestions = [
        { question: securityQuestions.securityQuestion1, answer: securityQuestions.securityAnswer1 },
        { question: securityQuestions.securityQuestion2, answer: securityQuestions.securityAnswer2 }
      ];
    }

    // Create or update the document
    const updatedDocument = await SecurityQuestion.findOneAndUpdate(
      { userId: userId },
      { $set: { securityQuestions, updatedAt: Date.now() } },
      { new: true, upsert: true, runValidators: true }
    );

    res.status(200).json(updatedDocument);
  } catch (error) {
    res.status(400).json({ error: error.message });
  }
};

// Get all security questions for a user
exports.getSecurityQuestions = async (req, res) => {
  try {
    const userId = req.user._id;
    const questions = await SecurityQuestion.find({ userId: userId });
    res.status(200).json(questions);
  } catch (error) {
    res.status(400).json({ error: error.message });
  }
};

// Update a security question
exports.updateSecurityQuestion = async (req, res) => {
  try {
    const { question, answer } = req.body;
    const updatedQuestion = await SecurityQuestion.findByIdAndUpdate(
      req.params.id,
      { question, answer, updatedAt: Date.now() },
      { new: true, runValidators: true }
    );
    if (!updatedQuestion) return res.status(404).json({ error: 'Question not found' });
    res.status(200).json(updatedQuestion);
  } catch (error) {
    res.status(400).json({ error: error.message });
  }
};

// Delete a security question
exports.deleteSecurityQuestion = async (req, res) => {
  try {
    const deletedQuestion = await SecurityQuestion.findByIdAndDelete(req.params.id);
    if (!deletedQuestion) return res.status(404).json({ error: 'Question not found' });
    res.status(200).json({ message: 'Question deleted successfully' });
  } catch (error) {
    res.status(400).json({ error: error.message });
  }
};

File: controllers/security_question_controller.py
from fastapi import HTTPException, Request
from models import SecurityQuestion  # Assuming Beanie model
from typing import Dict, Any, List

class SecurityQuestionController:
    @staticmethod
    async def add_or_update_security_questions(data: Dict[str, Any], user_id: str) -> Dict[str, Any]:
        """Add or update security questions for a user."""
        try:
            security_questions = data.get("securityQuestions", [])
            if not isinstance(security_questions, list) or not security_questions:
                security_questions = [
                    {"question": data.get("securityQuestion1"), "answer": data.get("securityAnswer1")},
                    {"question": data.get("securityQuestion2"), "answer": data.get("securityAnswer2")},
                ]

            updated_document = await SecurityQuestion.find_one(SecurityQuestion.user_id == user_id)
            if updated_document:
                updated_document.security_questions = security_questions
                updated_document.updated_at = datetime.now()
                await updated_document.save()
            else:
                updated_document = SecurityQuestion(
                    user_id=user_id,
                    security_questions=security_questions,
                    updated_at=datetime.now()
                )
                await updated_document.insert()

            return updated_document.dict()
        except Exception as e:
            raise HTTPException(status_code=400, detail=str(e))

    @staticmethod
    async def get_security_questions(user_id: str) -> Dict[str, Any]:
        """Retrieve all security questions for a user."""
        try:
            questions = await SecurityQuestion.find_one(SecurityQuestion.user_id == user_id)
            return {"questions": questions.dict() if questions else None}
        except Exception as e:
            raise HTTPException(status_code=400, detail=str(e))

    @staticmethod
    async def update_security_question(question_id: str, data: Dict[str, str]) -> Dict[str, Any]:
        """Update a specific security question."""
        try:
            question = await SecurityQuestion.get(question_id)
            if not question:
                raise HTTPException(status_code=404, detail="Question not found")

            question.question = data.get("question", question.question)
            question.answer = data.get("answer", question.answer)
            question.updated_at = datetime.now()
            await question.save()

            return question.dict()
        except Exception as e:
            raise HTTPException(status_code=400, detail=str(e))

    @staticmethod
    async def delete_security_question(question_id: str) -> Dict[str, str]:
        """Delete a specific security question."""
        try:
            question = await SecurityQuestion.get(question_id)
            if not question:
                raise HTTPException(status_code=404, detail="Question not found")

            await question.delete()
            return {"message": "Question deleted successfully"}
        except Exception as e:
            raise HTTPException(status_code=400, detail=str(e))